### PR TITLE
Adding a Documentation Generation Util

### PR DESCRIPTION
**Pull request by vibhatha**
# Description

This function generates the code documentation in markdown format. 

# Testing

## TODO

- [ ] Test cases for basic functionality
- [ ] Unit test each new function added
- [ ] Test cases for edge cases

# Documentation

## TODO

### CODE REVIEW

**Comment by vibhatha** [comment]
@aarondr77 the tests are failing, but I haven't added any tests yet.

**Review by aarondr77 — COMMENTED** [review]
Awesome work on this. I love a lot of the utilities that you built + how you were thinking about the user flow. 

A bit of cleanup that I think we should do and then we can get this merged!

File: mito-ai/src/utils/notebook.tsx [hunk 1]
```diff
@@ -80,3 +84,173 @@ export const highlightCodeCell = (notebookTracker: INotebookTracker, codeCellID:
...
+        // Add timeout to the request
+        const timeoutDuration = 30000; // 30 seconds
+        const timeoutPromise = new Promise((_, reject) => {
+            setTimeout(() => reject(new Error('Request timed out')), timeoutDuration);
+        });
+
+        const fetchPromise = requestAPI('mito_ai/completion', {
+            method: 'POST',
+            body: JSON.stringify({
+                messages: [{role: 'user', content: "Write markdown documentation for the following code:\n" + combinedCode}]
```
[the remark this answers is not in the document]
Comment (vibhatha) [reply mito-ai/src/utils/notebook.tsx:210]: @aarondr77 I agree, it would be better to refine this before we get this merged. I will work on it.

File: mito-ai/src/utils/notebook.tsx [hunk 2]
```diff
@@ -80,3 +84,173 @@ export const highlightCodeCell = (notebookTracker: INotebookTracker, codeCellID:
...
+        
+        if (error instanceof TimeoutError) {
+            errorMessage = new TimeoutError().message;
+        } else if (error.response?.status === 500) {
+            errorMessage = new ServerError().message;
+        } else if (error.message?.includes('Failed to fetch') || error.message?.includes('NetworkError')) {
+            errorMessage = new ConnectionError().message;
+        } else if (error instanceof OpenAIError) {
+            errorMessage = new OpenAIError().message;
+        }
```
[the remark this answers is not in the document]
Comment (vibhatha) [reply mito-ai/src/utils/notebook.tsx:238]: This is very naive for the moment, I think we could improve error handlers as we progress and keep a single wrapper to handle generic errors like calling a remote service and expecting a set of given failures and on top of that we can add other failure handlers depending on how custom they are compared to the base logic.

File: mito-ai/src/utils/notebook.tsx [hunk 3]
```diff
@@ -80,3 +84,173 @@ export const highlightCodeCell = (notebookTracker: INotebookTracker, codeCellID:
...
+    writeToCell(newCell.model, aiMessage);
+    NotebookActions.renderAllMarkdown(notebook.content);
+  } else {
+    console.error("New cell not found");
+  }
+}
+
+
+// Function to get combined code from selected cells
+export const getMarkdownDocumentation = async (notebookTracker: INotebookTracker): Promise<void> => {
```
[the remark this answers is not in the document]
Comment (vibhatha) [reply mito-ai/src/utils/notebook.tsx:172]: Right Good point. I will do that.

File: mito-ai/src/utils/notebook.tsx [hunk 4]
```diff
@@ -80,3 +84,173 @@ export const highlightCodeCell = (notebookTracker: INotebookTracker, codeCellID:
...
+    } catch (error: any) {
+        console.error('Error calling API:', error);
+        
+        let errorMessage = new UnknownError().message;
+        
+        if (error instanceof TimeoutError) {
+            errorMessage = new TimeoutError().message;
+        } else if (error.response?.status === 500) {
+            errorMessage = new ServerError().message;
+        } else if (error.message?.includes('Failed to fetch') || error.message?.includes('NetworkError')) {
```
[the remark this answers is not in the document]
Comment (vibhatha) [reply mito-ai/src/utils/notebook.tsx:234]: Right, let me check.

File: mito-ai/src/utils/notebook.tsx [hunk 5]
```diff
@@ -80,3 +84,173 @@ export const highlightCodeCell = (notebookTracker: INotebookTracker, codeCellID:
...
+    }
+  
+    for (let index = 0; index < notebook.widgets.length; index++) {
+      const cell = notebook.widgets[index];
+      if (cell && cell.model && cell.model.id === targetCellId) {
+        return index;
+      }
+    }
+  
+    return -1;
```
[the remark this answers is not in the document]
Comment (vibhatha) [reply mito-ai/src/utils/notebook.tsx:131]: Oh, right :) Thanks for this suggestion.

File: mito-ai/src/utils/notebook.tsx [hunk 6]
```diff
@@ -80,3 +84,173 @@ export const highlightCodeCell = (notebookTracker: INotebookTracker, codeCellID:
...
+}
+
+export function writeToCell(cell: ICellModel | undefined, code: string): void {
```
[the remark this answers is not in the document]
Comment (vibhatha) [reply mito-ai/src/utils/notebook.tsx:134]: Ah let me check and see how we can avoid the code duplication here.

File: mito-ai/src/utils/notebook.tsx [hunk 7]
```diff
@@ -1,6 +1,10 @@
-import { INotebookTracker } from '@jupyterlab/notebook';
-import { Cell } from '@jupyterlab/cells';
+import { INotebookTracker, NotebookPanel, Notebook, NotebookActions } from '@jupyterlab/notebook';
+import { Cell, ICellModel } from '@jupyterlab/cells';
 import { removeMarkdownCodeFormatting } from './strings';
+import { requestAPI } from './handler';
+import { ServerError, TimeoutError, ConnectionError, UnknownError, OpenAIError } from './errors';
+
+const LOADING_MARKDOWN = '> *`⏳ Generating documentation... please wait`*';
```
[the remark this answers is not in the document]
Comment (vibhatha) [reply mito-ai/src/utils/notebook.tsx:7]: You have a point there, but if we say 30 seconds, that is a guarantee, can we give that guarantee? 
Because this guarantee would depend on what underneath service guarantees us. Any idea about it?

**Comment by vibhatha** [comment]
@aarondr77 what is the best practice to deal with the yarn.lock file? Do we commit it?

File: mito-ai/src/utils/notebook.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (vibhatha) [reply mito-ai/src/utils/notebook.tsx:7]: I know my code has the time out, but that was actually a naive timeout I added without any reasoning.

File: mito-ai/src/Extensions/AiChat/AiChatPlugin.ts [hunk 8]
```diff
@@ -109,6 +111,31 @@ const AiChatPlugin: JupyterFrontEndPlugin<WidgetTracker> = {
       restorer.add(widget, 'mito_ai');
     }
 
+    app.commands.addCommand(COMMAND_MITO_AI_WRITE_MARKDOWN_DOC, {
```
[the remark this answers is not in the document]
Comment (vibhatha) [reply mito-ai/src/Extensions/AiChat/AiChatPlugin.ts:127]: Sure, let me remove it.

File: mito-ai/src/utils/notebook.tsx [hunk 9]
```diff
@@ -80,3 +84,173 @@ export const highlightCodeCell = (notebookTracker: INotebookTracker, codeCellID:
...
+            errorMessage = new ConnectionError().message;
+        } else if (error instanceof OpenAIError) {
+            errorMessage = new OpenAIError().message;
+        }
+
+        // Add error details for debugging (only in development)
+        const errorDetails = process.env.NODE_ENV === 'development' 
+            ? `\n\n<details><summary>Error Details</summary>\n\n\`\`\`\n${error.toString()}\n\`\`\`\n</details>`
+            : '';
+
```
[the remark this answers is not in the document]
Comment (vibhatha) [reply mito-ai/src/utils/notebook.tsx]: I think this is unnecessary, an artifact of code generation TBH. I don't think I intentionally added this. 
I tested using setting the env variable, it didn't work as expected, but better to explore for future usage.

File: mito-ai/src/utils/notebook.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (vibhatha) [reply mito-ai/src/utils/notebook.tsx]: @aarondr77 I think I am late in addressing this review. Because the `BasicPrompt.tsx` has been moved or renamed, Either way I improved my prompt a bit.

File: mito-ai/src/utils/notebook.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (vibhatha) [reply mito-ai/src/utils/notebook.tsx]: Done

File: mito-ai/src/utils/notebook.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (vibhatha) [reply mito-ai/src/utils/notebook.tsx]: Added a simple note.

File: mito-ai/src/utils/notebook.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (vibhatha) [reply mito-ai/src/utils/notebook.tsx]: Done

File: mito-ai/src/utils/notebook.tsx [hunk 10]
```diff
@@ -80,3 +84,173 @@ export const highlightCodeCell = (notebookTracker: INotebookTracker, codeCellID:
...
+    });
+    
+    try {
+        // Add timeout to the request
+        const timeoutDuration = 30000; // 30 seconds
+        const timeoutPromise = new Promise((_, reject) => {
+            setTimeout(() => reject(new Error('Request timed out')), timeoutDuration);
+        });
+
+        const fetchPromise = requestAPI('mito_ai/completion', {
```
[the remark this answers is not in the document]
Comment (vibhatha) [reply mito-ai/src/utils/notebook.tsx]: This is removed. Thanks for the feedback @aarondr77

File: mito-ai/src/Extensions/AiChat/AiChatPlugin.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (vibhatha) [reply mito-ai/src/Extensions/AiChat/AiChatPlugin.ts]: Done
